Remove depr hsa_code_object_t - find ISA from ELF

aaronenyeshi opened the pull request:
Required to use new hsa_code_object_reader_t for previous usage of hsa_code_object_t. Remove deprecated functions hsa_code_object_deserialize and hsa_executable_load_code_object. Also isCompatible ISA should find the target from ELF header to make decision. Remove code_object from ISA verification.

aaronenyeshi committed:
Remove depr hsa_code_object_t - find ISA from ELF

Required to use new hsa_code_object_reader_t for previous usage of hsa_code_object_t. Remove deprecated functions hsa_code_object_deserialize and hsa_executable_load_code_object. Also isCompatible ISA should find the target from ELF header to make decision. Remove code_object from ISA verification.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -2675,24 +2698,32 @@ class HSADevice final : public KalmarDevice
+        switch(MACH) {
+            case 0x023 : triple.append("701"); break;
+            case 0x02a : triple.append("803"); break;
+            case 0x02c : triple.append("900"); break;
```
@aaronenyeshi Can we avoid hard-coding ISA version here? We need to reduce future efforts required to support another new gfxXYZ.

kzhuravl replied:
@aaronenyeshi , those numbers are defined in here: https://github.com/RadeonOpenCompute/llvm/blob/0cd1006f4841272c27c95618acca4f0221b588fb/include/llvm/BinaryFormat/ELF.h#L691

aaronenyeshi replied:
I can't seem to include llvm/include/llvm/BinaryFormat/ELF.h without it conflicting (compile error) with /usr/include/elf.h.

aaronenyeshi replied:
@whchung I know there are many places we have target number such as lib/clamp-device, lit tests, Cmakes, clang driver. Maybe I should expose supported gfxXYZ in clang, and use the structure from there?

whchung replied:
@aaronenyeshi / @kzhuravl compared with including LLVM headers I think it'd be better to do this within clang. Not sure how easy would it be though.

kzhuravl replied:
Which things are you using from /usr/include/elf.h?

kzhuravl replied:
Can you switch from /usr/include/elf.h to llvm's ELF.h?

aaronenyeshi replied:
@AlexVlx Do you know if we can easily make the switch from /usr/include/elf.h to llvm's ELF.h?

scchan replied:
it looks like .note section in the elf object contains the gfx string, can we get it from there?

kzhuravl replied:
I highly do not recommend that. .note entry, which contains the gfx string, is going away in ROCM2.0.

scchan replied:
@aaronenyeshi where is /usr/include/elf.h needed?

aaronenyeshi replied:
@scchan I think in lib/hsa/mcwamp_hsa.cpp, there are many instances of `using namespace ELFIO;` and we use elf.h inside hc2/external/elfio/elfio_header.hpp

kzhuravl replied:
`+` @t-tye 

Maybe a long-term goal, but: should not we switch from using elfio to llvm infrastructure?

Not sure what to do in short-term... Should you redefine flags?

aaronenyeshi replied:
We've decided to mirror the llvm's ELF.h instead of using internal llvm header. It'll be easy to maintain instead of following LLVM's moving header.

scchan replied:
@aaronenyeshi  I don't see where we include elf.h in elfio_header.hpp?

@kzhuravl it doesn't seem that llvm's elf provides an API for parsing the elf object or iterating through the symbols/sections, which elfio provides

kzhuravl replied:
@scchan , I do not think I agree, take a look at:
https://github.com/llvm-mirror/llvm/blob/master/include/llvm/Object/ELF.h#L52
https://github.com/llvm-mirror/llvm/blob/master/include/llvm/Object/ELFObjectFile.h#L201

aaronenyeshi committed:
[ELF] Adding AMDGPU specific ELF enums to ELFIO

Adding amdgpu MACH e_flags into hcc ELFIO infrastructure in hc namespace. This will mirror LLVM's flags for amdgpu.

aaronenyeshi commented:
@kzhuravl How is this change to mirror llvm's elf header inside of our ELFIO. There shouldn't be frequent updates to AMDGPU e_flags I believe.

kzhuravl commented:
@aaronenyeshi  They should not be frequently updated. Those LGTM, but @scchan or @whchung  should approve changes.

aaronenyeshi commented:
@whchung Do you think an unordered_map could work? Maybe I'll associate in this way: 
umap[EF_AMDGPU_MACH_AMDGCN_GFX900] = "gfx900" ? I will create a new commit for you to take a look here

AlexVlx commented:
@aaronenyeshi you don't need a map, the enum is descriptive and you have the switch already. If something changes WRT the map from enum to string you still need to update something, so you're not saving much by using a map here.

aaronenyeshi committed:
Add target gfx906 to isCompatibleKernel

AlexVlx reviewed (approved):
LGTM.

aaronenyeshi commented:
@whchung @scchan what do you think about this implementation?

scchan commented on hc2/external/elfio/elfio_amdgpu.hpp:
```diff
@@ -0,0 +1,89 @@
+
+namespace hc {
+
+// AMDGPU specific e_flags.
```
could you add a comment to document where these hex values are defined/coming from?

aaronenyeshi replied:
Good point, just added it.

aaronenyeshi committed:
Add comment for src e_flag MACH codes

whchung reviewed (approved):
The PR looks good to me.

For future reference we probably need a guide to enumerate places within `hcc` (frontend , backend , runtime , device libs) in case a new GCN ISA target is to be added.
